### PR TITLE
work in progress about fusion layer

### PR DESCRIPTION
**Pull request by kwonjihun-theori**
I attempted to implement the fusion layer. Although it seems to have been implemented plausibly, the result only generates meaningless tokens, and the logits values from model(**inputs) contain only NaN values.
As far as I know, memory usage should increase a little after fusing, but I haven't observed any increase in memory usage either.

It seems like something is wrong, but it might be a minor issue like when I implemented quantize before. Although it's still incomplete, I thought it might be helpful, so I'm submitting a PR.

I'm not sure if this process is possible, but I have enough hardware to conduct tests. If you modify the code, I can test it.

I would also like to contribute to this AutoAWQ project.

### CODE REVIEW

**Comment by TechxGenus** [comment]
Thanks for the commit. The framework is correct and should work after some modifications.

File: awq/modules/fused/block.py [hunk 1]
```diff
@@ -132,6 +132,72 @@ def forward(
...
+        self.hidden_size = hidden_size
+        self.norm_1 = norm_1.to(dev)
+        self.attn = QuantAttentionFused(
+            self.hidden_size,
+            self.n_heads,
+            self.n_kv_heads,
+            qkv_layer,
+            o_proj,
+            dev=dev,
+            max_seq_len=max_seq_len,
```
Comment (TechxGenus) [inline awq/modules/fused/block.py:170]: Its RoPE implementation is different from Llama's. Require some modifications in `attn.py` to get everything working properly.

File: awq/models/cohere.py [hunk 2]
```diff
@@ -69,3 +75,57 @@ def get_layers_for_scaling(
...
+        module: OldCohereDecoderLayer
+        for module in tqdm.tqdm(self.model.model.layers, desc="Fusing layers..."):
+            device = next(iter(module.state_dict().values())).device
+            qkv = fuse_qkv(
+                module,
+                module.self_attn.q_proj,
+                module.self_attn.k_proj,
+                module.self_attn.v_proj,
+            )
+            norm_1 = FasterTransformerRMSNorm(
```
Comment (TechxGenus) [inline awq/models/cohere.py]: It use normal layernorm
```
norm_1 = module.input_layernorm
```

File: awq/modules/fused/block.py [hunk 3]
```diff
@@ -132,6 +132,72 @@ def forward(
...
+    ):
+        norm_out = self.norm_1(hidden_states)
+        attn_output, _, past_key_value = self.attn.forward(
+            hidden_states=norm_out,
+            past_key_value=past_key_value,
+            attention_mask=attention_mask,
+        )
+
+        h = hidden_states.to(attn_output.device) + attn_output
+        out = h + self.mlp.forward(h)
```
Comment (TechxGenus) [inline awq/modules/fused/block.py]:
```
out = h + self.mlp.forward(norm_out)
```

File: awq/modules/fused/block.py (hunk 1 above)
Comment (kwonjihun-theori) [reply awq/modules/fused/block.py:170]: ok others edit is done.

I will edit attn.py!

Comment (kwonjihun-theori) [reply awq/modules/fused/block.py:170]: I think this is too difficult task for me.

First, I checked the difference in the implementation of rotary posembedding between cohere and llama.

llama is just torch.cat after frequs operation, and cohere is torch.repeat_interleave.

Also, when llama proceeds with rotary_half
x1 = x[..., : x.shape[-1] // 2]
x2 = x[..., x.shape[-1] // 2 :]

to proceed,

cohere is
x1 = x[...,::2]
x2 = x[..., 1::2]

There is a difference in that.

However, I'm not sure how this should work with RoPE in awq.

Can I get any ideas or hints?

Comment (TechxGenus) [reply awq/modules/fused/block.py:170]: Aha, I think using a staggered rope should be fine. Now I have device to modify it.

**Comment by TechxGenus** [comment]
Thanks!
